module/state_synchronization/indexer/indexer_core.go

```diff
@@ -0,0 +1,174 @@
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+func (c *IndexerCore) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := c.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// This method shouldn't be used concurrently.
+// Expected errors:
+// - storage.ErrNotFound if the block for execution data was not found
+func (c *IndexerCore) IndexBlockData(data *execution_data.BlockExecutionDataEntity) error {
+	block, err := c.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
+	}
+
+	lg := c.log.With().
+		Hex("block_id", logging.ID(data.BlockID)).
+		Uint64("height", block.Height).
+		Logger()
+
+	lg.Debug().Msgf("indexing new block")
+
+	// the height we are indexing must be exactly one bigger or same as the latest height indexed from the storage
+	latest := c.registers.LatestHeight()
+	if block.Height != latest+1 && block.Height != latest {
+		return fmt.Errorf("must store registers with the next height %d, but got %d", latest+1, block.Height)
+	}
+	// allow rerunning the indexer for same height since we are fetching height from register storage, but there are other storages
+	// for indexing resources which might fail to update the values, so this enables rerunning and reindexing those resources
+	if block.Height == latest {
+		lg.Warn().Msg("reindexing block data")
+	}
+
+	// concurrently process indexing of block data
+	g := errgroup.Group{}
+
+	g.Go(func() error {
+		events := make([]flow.Event, 0)
+		for _, chunk := range data.ChunkExecutionDatas {
+			events = append(events, chunk.Events...)
+		}
+
+		err := c.indexEvents(data.BlockID, events)
+		if err != nil {
+			return fmt.Errorf("could not index events at height %d: %w", block.Height, err)
+		}
+		return nil
+	})
+
+	g.Go(func() error {
+		updates := make([]*ledger.TrieUpdate, 0)
+
+		for _, chunk := range data.ChunkExecutionDatas {
+			if chunk.TrieUpdate != nil {
+				updates = append(updates, chunk.TrieUpdate)
+			}
+		}
+
+		// we are iterating all the registers and overwrite any existing register at the same path
+		// this will make sure if we have multiple register changes only the last change will get persisted
+		// if block has two chucks:
+		// first chunk updates: { X: 1, Y: 2 }
+		// second chunk updates: { X: 2 }
+		// then we should persist only {X: 2: Y: 2}
+		payloads := make(map[ledger.Path]*ledger.Payload)
+		for _, update := range updates {
+			for i, path := range update.Paths {
+				payloads[path] = update.Payloads[i]
+			}
+		}
+
+		err = c.indexRegisters(maps.Values(payloads), block.Height)
```

Comment:
All we need is to iterate through all the payloads and create a slice of register entries to be passed to the storage API. Since the payloads has already been deduplicated, there is no difference between iterating through a map or a slice, That's why it's unnecessary to turn the map of payloads into a slice of payloads with `map.Values`

```suggestion
		err = c.indexRegisters(payloads, block.Height)
```